File: ACT_SASEP-master/SkySpace/JavaSky/src/main/java/com/skyspace/client/SkyClient.java
package com.skyspace.client;

import com.skyspace.element.Item;
import org.apache.commons.codec.binary.Base64;
import org.apache.commons.codec.digest.DigestUtils;
import org.apache.commons.lang3.StringUtils;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestTemplate;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Created by jianghaiting on 14/12/26.
 */
public class SkyClient {

    public static String SKY_SERVER = "http://127.0.0.1:9000";
    public static String WRITE_URL = "/skyentry/write";
    public static String READ_URL = "/skyentry/read";
    public static String TAKE_URL = "/skyentry/take";

    RestTemplate restTemplate = new RestTemplate();

    public String write(String... tuple) {
        String content = StringUtils.join(tuple, ",");
        String content_id = DigestUtils.sha1Hex(String.valueOf(Math.random()));
        MultiValueMap<String, Object> data = new LinkedMultiValueMap<String, Object>();
        data.add("content", content + "," + content_id);
        data.add("type", -1);
        data.add("expire", 3000);
        restTemplate.postForObject(SKY_SERVER + WRITE_URL, data, String.class);
        return content_id;
    }

    public List<String> read(String... template) {
        return readOrTake(SKY_SERVER + READ_URL, template);
    }

    public List<String> take(String... template) {
        return readOrTake(SKY_SERVER + TAKE_URL, template);
    }

    private List<String> readOrTake(String requestUrl, String[] template) {
        String content = StringUtils.join(template, ",");
        MultiValueMap<String, Object> data = new LinkedMultiValueMap<String, Object>();
        data.add("content", content);
        data.add("isMulti", false);
        data.add("timeout", 500);
        Item[] items = restTemplate.postForObject(requestUrl, data, Item[].class);

        if (items != null && items.length == 1) {
            String content1 = items[0].getContent();
            if (StringUtils.isBlank(content1)) {
                return null;
            } else {
                return Arrays.asList(content1.split(","));
            }
        } else {
            return null;
        }
    }

    private List<Object> decodeContent(String content) {
        String[] contents = content.split(",");
        List<Object> objs = new ArrayList<Object>();
        for (String c : contents) {
            String decodedContent = new String(Base64.decodeBase64(c));
            objs.add(decodedContent);
        }
        return objs;
    }

    private String encodeContent(Object[] contentItems) {
        List<String> encodedContentString = new ArrayList<String>();
        for (Object item : contentItems) {
            String str = item.toString();
            if (StringUtils.equals(str, "?")) {
                encodedContentString.add(str);
            } else {
                encodedContentString.add(Base64.encodeBase64String(str.getBytes()));
            }
        }
        return StringUtils.join(encodedContentString, ",");
    }


}

File: ACT_SASEP-master/AeroAnt/ResPool/sky_client.py
import logging

__author__ = 'jason'
import requests
import hashlib
import json
import base64
import random

SKY_SERVER = "http://127.0.0.1:9000"
WRITE_URL = "/skyentry/write"
READ_URL = "/skyentry/read"
TAKE_URL = "/skyentry/take"
RESET_URL = "/skyentry/reset"


def log():
    return logging.getLogger(__name__)


def __handle_result(r, is_multi, return_id):
    if r.status_code != 200:
        log().error("request failed, status code = %d", r.status_code)
        return
    ret = r.json()
    if not is_multi:
        if len(ret) != 1:
            if return_id:
                return None, None
            else:
                return None
        else:
            ret = ret[0]
            content = ret["content"].split(",")
            if return_id:
                return content[0:-1], content[-1]
            else:
                return content
    else:
        print "not implement yet"
        return None


def write(tuple, type=-1, expire=5000, fill_content_id=False):
    content, cid = get_content(tuple, fill_content_id)
    param = {
        "content": content,
        "type": type,
        "expire": expire
    }
    log().debug("SKY_CLIENT[write],data=%s", param)
    requests.post(SKY_SERVER + WRITE_URL, data=param)
    return cid


def read(template, is_multi=False, timeout=500, return_id=False):
    content, cid = get_content(template, False)
    param = {
        "content": content,
        "isMulti": is_multi,
        "timeout": timeout
    }
    log().debug("SKY_CLIENT[read],data=%s", param)
    r = requests.post(SKY_SERVER + READ_URL, data=param)
    return __handle_result(r, is_multi, return_id)


def take(template, is_multi=False, timeout=5000, return_id=False):
    content, cid = get_content(template, False)
    param = {
        "content": content,
        "isMulti": is_multi,
        "timeout": timeout
    }
    log().debug("SKY_CLIENT[take],data=%s", param)
    r = requests.post(SKY_SERVER + TAKE_URL, data=param)
    return __handle_result(r, is_multi, return_id)


def get_content(item_list, fill_content_id=True):
    """
    this will append content_id in result!
    :param item_list:
    :param return_id:
    :return: content , content_id
    """
    #print "get_content:", item_list, return_id
    # content = ",".join(map(__encode, item_list))

    content = ",".join(item_list or [])
    content_id = hashlib.sha1(content + str(random.random())).hexdigest()
    if fill_content_id:
        content = content + "," + content_id
    return content, content_id


def reset():
    r = requests.post(SKY_SERVER + RESET_URL)
    return r.status_code == 200